### PR TITLE
Fortran ordering, array of arrays, nested arrays

### LINKED ISSUE #1258: Fortran arrays are not correctly wrapped
### Describe the bug
Fortran arrays are not correctly wrapped, so the result is transposed

### To Reproduce
Provide code to reproduce the behavior:
```python
def f():
    import numpy as np
    a = np.array([[1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16]], order="F")
    return a
```
if we compile this function and run it in the following code:
```python
import tmp

x = tmp.f()
print(x)
```

### Error details
Python output:
```bash
[[ 1  2  4  8 16]
 [ 1  2  4  8 16]
 [ 1  2  4  8 16]
 [ 1  2  4  8 16]
 [ 1  2  4  8 16]]
```

Pyccelised output:
```bash
[[ 1  1  1  1  1]
 [ 2  2  2  2  2]
 [ 4  4  4  4  4]
 [ 8  8  8  8  8]
 [16 16 16 16 16]]
```

### Expected behavior
The array shouldn't be transposed

### Language
Fortran

### Additional context
Identified in PR https://github.com/pyccel/pyccel/pull/1256
Blocked by #1280

### PR DESCRIPTION
**Pull request by ohachim**
Implements the function `array_copy_data` and updates the logic to ensure that fortran order memory layouts are correctly handled.

Implements fortran order memory layout array creation (`np.array`).
Implements nested array creation.

Fixes https://github.com/pyccel/pyccel/issues/752
Fixes https://github.com/pyccel/pyccel/issues/1179
Fixes https://github.com/pyccel/pyccel/issues/682
Fixes #1258

### CODE REVIEW

File: pyccel/stdlib/ndarrays/ndarrays.c [hunk 1]
```diff
@@ -371,5 +441,70 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
...
+    if (!src.is_view && dest->order == src.order
+        && (src.order == order_c
+            || (src.order == order_f && is_same_shape(*dest, src))))
+        memcpy(d + offset * dest->type_size, s, src.buffer_size);
+    else
+    {
+        for (int32_t element_num = 0; element_num < src.length; ++element_num)
+        {
+            memcpy(d + ((element_index(*dest, element_num, dest->nd) + offset) * dest->type_size),
+                s + (element_index(src, element_num, src.nd) * src.type_size), src.type_size);
```
Comment (EmilyBourne) [inline pyccel/stdlib/ndarrays/ndarrays.c:507]: I presume this is written like this as it was the only way to avoid a huge if/else block?

File: tests/ndarrays/test_ndarrays.c [hunk 2]
```diff
@@ -178,6 +178,35 @@ int32_t test_indexing_int64(void)
...
+    int32_t c_index;
+    int64_t value;
+    int64_t c_value;
+
+    x = array_create(2, m_1_shape, nd_int64, false, order_f);
+    memcpy(x.raw_data, m_1, x.buffer_size);
+    // testing the index [3, 2]
+    index = 3 * x.strides[0] + 2 * x.strides[1];
+    c_index = 13;
+    my_assert(index , c_index, "testing the strides");
```
Comment (EmilyBourne) [inline tests/ndarrays/test_ndarrays.c:200]: This tests the strides not the index

File: tests/ndarrays/test_ndarrays.c [hunk 3]
```diff
@@ -265,6 +353,35 @@ int32_t test_indexing_int8(void)
...
+    int8_t value;
+    int8_t c_value;
+
+    x = array_create(2, m_1_shape, nd_int8, false, order_f);
+    memcpy(x.raw_data, m_1, x.buffer_size);
+    // testing the index [3, 2]
+    index = 3 * x.strides[0] + 2 * x.strides[1];
+    c_index = 13;
+    my_assert(index , c_index, "testing the strides");
+    my_assert(get_index(x, 3, 2) , c_index, "testing the indexing function");
```
Comment (EmilyBourne) [inline tests/ndarrays/test_ndarrays.c:376]: This testing is good, but as you need multiple functions anyway, why not test a few more indexes?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/ccode.py [hunk 4]
```diff
@@ -303,6 +303,28 @@ def _flatten_list(self, irregular_list):
...
+        flattened_list: list to search in
+
+        Return
+        ------
+            starting_consecutive_scalars: starting consecutive scalars
+        """
+
+        starting_consecutive_scalars = []
+        for i in flattened_list:
+            if not isinstance(i, Variable) or ((isinstance(i, Variable)) and i.rank == 0):
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py]: Yes it is wrong, I changed it to `if i.rank == 0`
https://github.com/pyccel/pyccel/pull/1280/commits/de588d6a95a953c7e1f85f6aa57bf58c79d737cb

File: pyccel/codegen/printing/ccode.py [hunk 5]
```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
...
-            return '%s' % (cpy_data)
-        else :
-            if arg.rank > 1:
-                # flattening the args to use them in C initialization.
-                arg = self._flatten_list(arg)
-            arg = ', '.join(self._print(i) for i in arg)
-            dummy_array = "%s %s[] = {%s};\n" % (declare_dtype, dummy_array_name, arg)
-            cpy_data = "memcpy({0}.{2}, {1}, {0}.buffer_size);\n".format(self._print(lhs), dummy_array_name, dtype)
-            return  '%s%s' % (dummy_array, cpy_data)
+            return f"array_copy_data(&{lhs_name}, {self._print(arg)}, 0);\n"
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py:349]: I don't understand what you meant by optional
this piece of code runs for somthing like this
```python
b = np.array([1, 2, 3])
a = np.array(b)
```

File: pyccel/codegen/printing/ccode.py [hunk 6]
```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
...
+        order = lhs.order
+        rhs_dtype = self._print(rhs.dtype)
+        declare_dtype = self.find_in_dtype_registry(rhs_dtype, rhs.precision)
+        dtype = self.find_in_ndarray_type_registry(rhs_dtype, rhs.precision)
+        flattened_list = self._flatten_list(arg)
+        i = 0
+        operations = ""
+        if order == "F":
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py:357]: The list flattening?
```python
    d = 4
    b = numpy.array([1, 2, 3], order="F")
    a = numpy.array([b, [1, d, 3]])
    print(a)
```
prints
```
[[1 2 3]
[1 4 3]]
```
and this is the `a`'s memory layout: `[1][2][3][1][4][3]`

[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py:357]: `array_copy_data` takes cares of the order diffrences

File: pyccel/codegen/printing/ccode.py [hunk 7]
```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
...
+            nd = arg.rank
+            temp_array_name = self.scope.get_new_name('temp_array')
+            shape_dtype = self.find_in_dtype_registry('int', 8)
+            shape = ', '.join(self._print(elem) for elem in lhs.shape)
+            shape_Assign = "("+ shape_dtype +"[]){" + shape + "}"
+            temp_array_declaration = f"t_ndarray {temp_array_name} = " "{.shape = NULL};\n"
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py]: https://github.com/pyccel/pyccel/commit/1f2e3e6b3edc67ed1771b181dfb6de1c237bd203

File: pyccel/codegen/printing/ccode.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py:349]: But I can see how that using `ObjectAddress` is better!

File: pyccel/codegen/printing/ccode.py [hunk 8]
```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
...
+        and num_elements != 1:
+            offset_name = self.scope.get_new_name('offset')
+            operations += f"uint32_t {offset_name} = 0;\n"
+        else:
+            offset_name = ""
+        while i < num_elements:
+            if isinstance(flattened_list[i], Variable) and flattened_list[i].rank >= 1:
+                elem_name = self._print(flattened_list[i])
+                offset = offset_name if offset_name else "0"
+                operations += f"array_copy_data(&{copy_to}, {elem_name}, {offset});\n"
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py]: https://github.com/pyccel/pyccel/pull/1280/commits/990a7915987d3c8028cfa3236f3397676791ae3b

[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py]: I've been commenting the commits without waiting for the tests 🤦
I'll wait for them now

File: pyccel/codegen/printing/ccode.py [hunk 9]
```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
...
+        if order == "F":
+            nd = arg.rank
+            temp_array_name = self.scope.get_new_name('temp_array')
+            shape_dtype = self.find_in_dtype_registry('int', 8)
+            shape = ', '.join(self._print(elem) for elem in lhs.shape)
+            shape_Assign = "("+ shape_dtype +"[]){" + shape + "}"
+            temp_array_declaration = f"t_ndarray {temp_array_name} = {{.shape = NULL}};\n"
+            array_creation = f"{temp_array_name} = array_create({nd}, {shape_Assign}, {dtype}, {'false'}, {'order_c'});\n"
+            operations += temp_array_declaration + array_creation
+            copy_to = temp_array_name
```
[the remark this answers is not in the document]
Comment (aihya) [reply pyccel/codegen/printing/ccode.py:366]: Although, I suggest to you figure out a way to do this in the semantic part instead of here because you'll have a variable declaration followed by allocation repeated too many times, especially if you repeat the call to this function.

File: pyccel/codegen/printing/ccode.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: The equivalent with an optional would be something like:
```python
def f(b = None : 'int[:]'):
    if b is None:
        a = np.ones(4)
    else:
        a = np.array(b)
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: It's probably worth adding a test

File: pyccel/codegen/printing/ccode.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:366]: I disagree about moving it to the semantic stage. Things in the semantic stage are applied to all languages and there is no reason to have all this in a Fortran or python translation

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:366]: I'm also not sure how moving it to the semantic stage would avoid the problem you're describing

[the remark this answers is not in the document]
Comment (aihya) [reply pyccel/codegen/printing/ccode.py:366]: True, I was just considering how the code is going to be printed. Instead of having a declaration followed by an allocation multiple times, I thought it might be cleaner to have bunch of declarations and then the allocations.

[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py]: Using `Declare` and `Allocate` does seem like a good idea (https://github.com/pyccel/pyccel/pull/1280/commits/17bec94d8ea5d3659999ec4105e96e6ae6b8c867)
@aihya @EmilyBourne I used `status="unallocated"` for the `Allocate` instance, but I could not understand what it refers to from the docs
(why do we have to possibility to set it to `allocated` when the class itself is used to allocate)

File: pyccel/codegen/printing/ccode.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py]: I see!
ok

File: pyccel/codegen/printing/ccode.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (aihya) [reply pyccel/codegen/printing/ccode.py]: The status is used in `_print_Allocate` to signify whether the variable should be deallocated if it already exist in the memory.

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: > True, I was just considering how the code is going to be printed. Instead of having a declaration followed by an allocation multiple times, I thought it might be cleaner to have bunch of declarations and then the allocations.

I agree, but this can also be done from the `CCodePrinter`. You just need to insert the `Variable` into the scope so the `Declare` object is automatically created.

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: > why do we have to possibility to set it to allocated when the class itself is used to allocate)

@aihya is right, consider the 2 following examples:
```python
a = np.empty(4)       # Allocate('a', allocated = 'unallocated')
....
a = np.empty(5)       # Allocate('a', allocated = 'allocated')
```
and
```python
for i in range(5):
    a = np.empty(i+1)    # Allocate('a', allocated = 'unknown') (unallocated when i == 0, allocated after that)
```

File: tests/ndarrays/test_ndarrays.c (hunk 2 above)
Comment (ohachim) [reply tests/ndarrays/test_ndarrays.c:200]: Yes, I tried to copy how the other tests were done, the strides are tested anyway (the index function is tested on the following line), adding strides tests here seems productive imo, since strides and indexing are related

File: pyccel/stdlib/ndarrays/ndarrays.c (hunk 1 above)
Comment (ohachim) [reply pyccel/stdlib/ndarrays/ndarrays.c:507]: Yes, so that we can copy straight to `raw_data`

File: tests/ndarrays/test_ndarrays.c (hunk 2 above)
Comment (EmilyBourne) [reply tests/ndarrays/test_ndarrays.c:200]: My point is that the comment says "testing the index"

File: pyccel/codegen/printing/ccode.py [hunk 10]
```diff
@@ -313,31 +335,67 @@ def copy_NumpyArray_Data(self, expr):
...
+            operations += f"uint32_t {offset_name} = 0;\n"
+        else:
+            offset_name = ""
+        while i < num_elements:
+            if isinstance(flattened_list[i], Variable) and flattened_list[i].rank >= 1:
+                elem_name = self._print(flattened_list[i])
+                offset = offset_name if offset_name else "0"
+                operations += f"array_copy_data(&{copy_to}, {elem_name}, {offset});\n"
+                if i < num_elements - 1:
+                    operations += f"{offset_name} += {elem_name}.length;\n"
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py:379]: `PyccelArraySize` prints out `shape` in a given dimension
Something like `PyccelArraySize(Variable(e), 3)` would print ```e.shape[3]```, we would have to print this in a loop to do the shape multiplication?

[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/codegen/printing/ccode.py:379]: We're talking about the last line right? replacing `{elem_name}.length`

File: tests/ndarrays/test_ndarrays.c (hunk 2 above)
Comment (ohachim) [reply tests/ndarrays/test_ndarrays.c:200]: "testing the strides"? or do you mean that we should change the comment to "testing the index"?
```c  
    index = 3 * x.strides[0] + 2 * x.strides[1];
    c_index = 13;
    my_assert(index , c_index, "testing the strides"); // making sure that the strides are correct
    my_assert(get_index(x, 3, 2) , c_index, "testing the indexing function"); // checking that that 'get_index' is correct
```
The `get_index` function does use strides to calculate index anyway, so maybe it is counterproductive to check for them there, maybe we should just remove that line from all the tests
@bauom

File: pyccel/codegen/printing/ccode.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:379]: Sorry I meant `NumpyArraySize`

File: tests/ndarrays/test_ndarrays.c (hunk 2 above)
Comment (EmilyBourne) [reply tests/ndarrays/test_ndarrays.c:200]: Line 197. The start of this code snippet:
```
// testing the index [3, 2]
```

Comment (EmilyBourne) [reply tests/ndarrays/test_ndarrays.c:200]:
```suggestion
    // testing the strides for index [3, 2]
    index = 3 * x.strides[0] + 2 * x.strides[1];
    c_index = 13;
    my_assert(index , c_index, "testing the strides");
```

Comment (ohachim) [reply tests/ndarrays/test_ndarrays.c:200]: oh right! mb

File: pyccel/stdlib/ndarrays/ndarrays.c [hunk 11]
```diff
@@ -371,5 +441,70 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
...
         nd_shape[i] = np_shape[i];
     return nd_shape;
+}
 
+/*
+** takes an array containing the shape of an array 'shape', number of a certain dimension 'nd', and the number of the array's dimensions
+** returns the stride (number of single elements to jump in a dimension to get to this dimension's next element) of the 'nd`th dimension
+*/
+
+int get_dimension_stride(int64_t *shape, int nd, int max_nd)
```
[the remark this answers is not in the document]
Comment (ohachim) [reply pyccel/stdlib/ndarrays/ndarrays.c:451]: I agree! would `dimension_id` work or is it too wordy?

**Comment by EmilyBourne** [comment]
/bot run linux coverage docs pylint python_lint spelling

**Comment by EmilyBourne** [comment]
/bot run linux coverage docs pylint pyccel_lint spelling

**Comment by EmilyBourne** [comment]
/bot run linux coverage docs pylint pyccel_lint spelling

File: AUTHORS [hunk 12]
```diff
@@ -24,3 +24,5 @@ Contributors
 * Salah Eddine Mhah
 * Allaeddine Khalid
 * Anass Maach
+* Othmane Hachim
+* soukainahikma
```
Comment (EmilyBourne) [inline AUTHORS:28]: @soukainahikma @bauom Can we update this to the correct name?

**Comment by EmilyBourne** [comment]
I have no explanation for the MacOS failure here : https://github.com/pyccel/pyccel/actions/runs/5665623506
It seems to run fine on Linux and I don't have access to a mac so I cannot investigate.

**Review by EmilyBourne — APPROVED** [review]

**Comment by EmilyBourne** [comment]
@yguclu This should be merged after #1394 as there will be conflicts and #1394 is already ready to merge

**Review by EmilyBourne — APPROVED** [review]

**Review by EmilyBourne — APPROVED** [review]

**Review by EmilyBourne — APPROVED** [review]

**Review by EmilyBourne — APPROVED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -21,7 +24,7 @@ All notable changes to this project will be documented in this file.
 -   Pyccel-generated folder names are dependent on `PYTEST_XDIST_WORKER` when running with `pytest-xdist`.
 -   \[INTERNALS\] Add class object to class function call arguments.
 -   \[INTERNALS\] In `ast.numpyext` rename `Shape` as `NumpyShape`, `NumpyArraySize` as `NumpySize`
--   \[INTERNALS\] In `ast.internals` rename `NumpyArraySize` as `NumpySize`, create new `NumpyArraySize` w/out `index` argument
+-   \[INTERNALS\] In `ast.internals` rename `PyccelArraySize` as `PyccelArraySizeElement`, create new `PyccelArraySize` w/out `index` argument
```
Comment (yguclu) [inline CHANGELOG.md:27]: Good catch 😅

**Review by EmilyBourne — APPROVED** [review]

**Comment by EmilyBourne** [comment]
I'm not sure what is wrong with the API but it is seeing my approval as a comment :shrug:

**Review by yguclu — CHANGES_REQUESTED** [review]
Good job!

I have a couple of questions and remarks.

File: pyccel/stdlib/ndarrays/ndarrays.c [hunk 14]
```diff
@@ -7,18 +7,76 @@
...
+                printf("[%lf]", nd.nd_double[i]);
+                break;
+            case nd_bool:
+                printf("[%d]", nd.nd_bool[i]);
+                break;
+            case nd_cfloat:
+            {
+                double real = creal(nd.nd_cfloat[i]);
+                double imag = cimag(nd.nd_cfloat[i]);
+                printf("[%lf%s%lfj]", real, imag >= 0 ? "+" : "", imag);
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/stdlib/ndarrays/ndarrays.c:51]: E.g.
if `imag=3` this will print with the format `{real}+{imag}` e.g. 3+5j
if `imag=-3` this will print with the format `{real}{imag}` e.g. 3-5j

File: tests/epyccel/modules/arrays.py [hunk 15]
```diff
@@ -397,7 +397,141 @@ def array_real_4d_C_array_initialization(x, y, a):
     from numpy import array
     tmp      = array([x, y], dtype='float')
     a[:,:,:,:] = tmp[:,:,:,:]
+##==============================================================================
+## TEST NESTED ARRAYS INITIALIZATION WITH ORDER C
+##==============================================================================
 
+@types('real[:,:,:]','real[:,:]', 'real[:,:]', 'real[:,:,:,:]')
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/modules/arrays.py:404]: This has been discussed multiple times. It is not incompatible with pure python if the import is correctly included `from pyccel.decorators import types`. It can be replaced by `@template` to some degree, but it gives more flexibility to limit the number of cases generated. E.g:
```
@types('float[:]','int[:]')
@types('float[:,:]','int[:,:]')
@types('float[:,:,:]','int[:,:,:]')
def f(a, b):
    pass
```
which describes 3 cases where arrays have the same number of dimensions. The equivalent with templates would lead to 9 cases including cases where arrays have different numbers of dimensions which could potentially be problematic:
```
@template('T', ['float[:]','float[:,:]','float[:,:,:]'])
@template('T', ['int[:]','int[:,:]','int[:,:,:]'])
def f(a, b):
    pass
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/modules/arrays.py:404]: We would have significantly fewer examples if we went through and changed the syntax in these files as many new tests are created by copying from a nearby test

File: tests/epyccel/test_arrays.py [hunk 16]
```diff
@@ -3679,9 +3703,168 @@ def test_multiple_np_linspace(language):
...
+## TEST ARRAY VIEW STEPS ARRAY INITIALIZATION ORDER F 2D
 ##==============================================================================
-#
+
+@pytest.mark.parametrize( 'language', [
+        pytest.param("c", marks = pytest.mark.c),
+        pytest.param("fortran", marks = [pytest.mark.skip(reason="fortran bug, issue #1241"),
+                                        pytest.mark.fortran])
+    ]
+)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3831]: Fixed 4e8e7f63413441e778d95223b46e786d8084c23d

File: tests/epyccel/test_arrays.py [hunk 17]
```diff
@@ -3679,9 +3703,168 @@ def test_multiple_np_linspace(language):
...
+    f1 = arrays.array_view_steps_F_2D_1
+    f2 = epyccel(f1, language = language)
+    assert np.array_equal(f1(a), f2(a))
+
+@pytest.mark.parametrize( 'language', [
+        pytest.param("c", marks = pytest.mark.c),
+        pytest.param("fortran", marks = [pytest.mark.skip(reason="fortran bug, issue #1241"),
+                                        pytest.mark.fortran])
+    ]
+)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3845]: Fixed 4e8e7f63413441e778d95223b46e786d8084c23d

File: tests/epyccel/test_arrays.py [hunk 18]
```diff
@@ -3679,9 +3703,168 @@ def test_multiple_np_linspace(language):
...
+    f1 = arrays.array_view_steps_F_2D_2
+    f2 = epyccel(f1, language = language)
+    assert np.array_equal(f1(a), f2(a))
+
+@pytest.mark.parametrize( 'language', [
+        pytest.param("c", marks = pytest.mark.c),
+        pytest.param("fortran", marks = [pytest.mark.skip(reason="fortran bug, issue #1241"),
+                                        pytest.mark.fortran])
+    ]
+)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3859]: Fixed 4e8e7f63413441e778d95223b46e786d8084c23d

File: tests/epyccel/test_arrays.py [hunk 19]
```diff
@@ -3679,9 +3703,168 @@ def test_multiple_np_linspace(language):
...
 
 ##==============================================================================
-## CLEAN UP GENERATED FILES AFTER RUNNING TESTS
+## TEST NESTED ARRAYS INITIALIZATION WITH ORDER F
+##==============================================================================
+
+def test_array_real_nested_F_array_initialization(language):
+
+
+    f1 = arrays.array_real_nested_F_array_initialization
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3712]: Fixed 4e8e7f63413441e778d95223b46e786d8084c23d

File: pyccel/stdlib/ndarrays/ndarrays.h [hunk 20]
```diff
@@ -73,6 +73,12 @@ enum e_types
         nd_cdouble  = 15
 };
 
+enum e_order
+{
+    order_f,
+    order_c,
+};
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/stdlib/ndarrays/ndarrays.h:80]: Fixed 68ac1e95bdd0dd55624f8e954a2cbd2053086ce7

File: pyccel/stdlib/ndarrays/ndarrays.c [hunk 21]
```diff
@@ -363,7 +432,74 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
     for (int i = 0; i < nd; i++)
         nd_shape[i] = np_shape[i];
     return nd_shape;
+}
+
+/*
+** takes an array containing the shape of an array 'shape', number of a certain dimension 'nd', and the number of the array's dimensions
+** returns the stride (number of single elements to jump in a dimension to get to this dimension's next element) of the 'nd`th dimension
+*/
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/stdlib/ndarrays/ndarrays.c:440]: Fixed 1a3f621c4ed356d23ddf24a4c05c32d511d12e31

File: pyccel/stdlib/ndarrays/ndarrays.c [hunk 22]
```diff
@@ -363,7 +432,74 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
...
+        product *= shape[i];
+    return (product);
+}
+
+/*
+**  Takes an array needed to do the calculations
+**  An element number, representing an element's index if it were in a flattened (order_c/row major) array
+**  nd, representing the number of dimensions
+**  returns the element's index depending on its required memory layout (order_f/column major or order_c/row major)
+*/
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/stdlib/ndarrays/ndarrays.c:456]: Fixed 1a3f621c4ed356d23ddf24a4c05c32d511d12e31

File: pyccel/stdlib/ndarrays/ndarrays.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/stdlib/ndarrays/ndarrays.c:51]: Is there not a signed print format in C? It should be `%+lf` if I am not mistaken:
```C
printf("[%lf%+lfj]", real, imag)
```

File: tests/epyccel/modules/arrays.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply tests/epyccel/modules/arrays.py]: I meant just the ones in this PR. We have to start from somewhere 😉

[the remark this answers is not in the document]
Comment (yguclu) [reply tests/epyccel/modules/arrays.py]: > This has been discussed multiple times. It is not incompatible with pure python if the import is correctly included `from pyccel.decorators import types`. It can be replaced by `@template` to some degree, but it gives more flexibility to limit the number of cases generated. E.g:
> 
> ```
> @types('float[:]','int[:]')
> @types('float[:,:]','int[:,:]')
> @types('float[:,:,:]','int[:,:,:]')
> def f(a, b):
>     pass
> ```
> 
> which describes 3 cases where arrays have the same number of dimensions. The equivalent with templates would lead to 9 cases including cases where arrays have different numbers of dimensions which could potentially be problematic:
> 
> ```
> @template('T', ['float[:]','float[:,:]','float[:,:,:]'])
> @template('T', ['int[:]','int[:,:]','int[:,:,:]'])
> def f(a, b):
>     pass
> ```

True, there are some use cases for the `@types` annotation, but in most instances they could be replaced with either the `@elemental` annotation or the "partial templates", once we get to finally implement them.

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/modules/arrays.py]: > I meant just the ones in this PR. We have to start from somewhere :wink:

There are still quite a lot of tests in this PR so it will take more time to correct just these ones than it would to write a script to correct them all. I will create a PR to remove unnecessary `@types`.

File: pyccel/stdlib/ndarrays/ndarrays.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/stdlib/ndarrays/ndarrays.c:51]: With the signed print format that I mentioned above, there should be no need for using the conditional ternary operator. Can you please check?

**Review by yguclu — APPROVED** [review]
